File: sdk/javascript/src/crypto.js
/**
 * OpenADP Cryptographic Operations
 * 
 * This module provides all cryptographic primitives used by OpenADP:
 * - Ed25519 elliptic curve operations
 * - Point compression/decompression
 * - HKDF key derivation
 * - Shamir secret sharing
 * - Authentication code generation
 */

import crypto from 'crypto';
import forge from 'node-forge';

// Ed25519 curve parameters
const P = 2n ** 255n - 19n;  // Field prime
const Q = 2n ** 252n + 27742317777372353535851937790883648493n;  // Curve order
const D = 37095705934669439343138083508754565189542113879843219016388785533085940283555n;  // Curve parameter

// Base point G coordinates
const GX = 15112221349535400772501151409588531511454012693041857206046113283949847762202n;
const GY = 46316835694926478169428394003475163141307993866256225615783033603165251855960n;

/**
 * Point on Ed25519 curve in extended coordinates (X, Y, Z, T)
 */
class Point4D {
    constructor(x = 0n, y = 1n, z = 1n, t = 0n) {
        this.x = BigInt(x);
        this.y = BigInt(y);
        this.z = BigInt(z);
        this.t = BigInt(t);
    }

    /**
     * Convert to affine coordinates (x, y)
     */
    toAffine() {
        if (this.z === 0n) {
            throw new Error('Point at infinity cannot be converted to affine coordinates');
        }
        const zInv = modInverse(this.z, P);
        return new Point2D(mod(this.x * zInv, P), mod(this.y * zInv, P));
    }

    /**
     * Point addition in extended coordinates
     */
    add(other) {
        const a = mod((this.y - this.x) * (other.y - other.x), P);
        const b = mod((this.y + this.x) * (other.y + other.x), P);
        const c = mod(2n * this.t * other.t * D, P);
        const d = mod(2n * this.z * other.z, P);
        const e = mod(b - a, P);
        const f = mod(d - c, P);
        const g = mod(d + c, P);
        const h = mod(b + a, P);

        return new Point4D(
            mod(e * f, P),
            mod(g * h, P),
            mod(f * g, P),
            mod(e * h, P)
        );
    }

    /**
     * Point doubling in extended coordinates
     */
    double() {
        // Use point addition with itself for consistency
        return this.add(this);
    }

    /**
     * Scalar multiplication using double-and-add
     */
    multiply(scalar) {
        let result = new Point4D(0n, 1n, 1n, 0n);  // Identity element
        let base = new Point4D(this.x, this.y, this.z, this.t);
        let k = BigInt(scalar);

        while (k > 0n) {
            if (k & 1n) {
                result = result.add(base);
            }
            base = base.double();
            k >>= 1n;
        }

        return result;
    }

    /**
     * Check if point is the identity element
     */
    isIdentity() {
        const affine = this.toAffine();
        return affine.x === 0n && affine.y === 1n;
    }

    /**
     * Check if point is valid on the curve
     */
    isValid() {
        try {
            const affine = this.toAffine();
            return affine.isValid();
        } catch {
            return false;
        }
    }
}

/**
 * Point on Ed25519 curve in affine coordinates (x, y)
 */
class Point2D {
    constructor(x = 0n, y = 1n) {
        this.x = BigInt(x);
        this.y = BigInt(y);
    }

    /**
     * Convert to extended coordinates
     */
    toExtended() {
        return new Point4D(this.x, this.y, 1n, mod(this.x * this.y, P));
    }

    /**
     * Check if point is valid on the curve: -x^2 + y^2 = 1 + d*x^2*y^2
     */
    isValid() {
        const x2 = mod(this.x * this.x, P);
        const y2 = mod(this.y * this.y, P);
        const left = mod(y2 - x2, P);
        const right = mod(1n + D * x2 * y2, P);
        return left === right;
    }

    /**
     * Compress point to 32 bytes (Ed25519 format)
     */
    compress() {
        const bytes = new Uint8Array(32);
        
        // Encode y coordinate (little-endian)
        let y = this.y;
        for (let i = 0; i < 32; i++) {
            bytes[i] = Number(y & 0xFFn);
            y >>= 8n;
        }
        
        // Set sign bit based on x coordinate parity
        if (this.x & 1n) {
            bytes[31] |= 0x80;
        }
        
        return bytes;
    }

    /**
     * Decompress point from 32 bytes (Ed25519 format)
     */
    static decompress(bytes) {
        if (bytes.length !== 32) {
            throw new Error('Invalid compressed point length');
        }
        
        // Extract sign bit
        const signBit = (bytes[31] & 0x80) !== 0;
        
        // Extract y coordinate (clear sign bit)
        const yBytes = new Uint8Array(bytes);
        yBytes[31] &= 0x7F;
        
        let y = 0n;
        for (let i = 31; i >= 0; i--) {
            y = (y << 8n) + BigInt(yBytes[i]);
        }
        
        if (y >= P) {
            throw new Error('Invalid y coordinate');
        }
        
        // Solve for x: x^2 = (y^2 - 1) / (d*y^2 + 1)
        const y2 = mod(y * y, P);
        const numerator = mod(y2 - 1n, P);
        const denominator = mod(D * y2 + 1n, P);
        const x2 = mod(numerator * modInverse(denominator, P), P);
        
        // Compute square root
        let x = modSqrt(x2, P);
        if (x === null) {
            throw new Error('Point not on curve');
        }
        
        // Adjust sign
        if ((x & 1n) !== BigInt(signBit ? 1 : 0)) {
            x = mod(-x, P);
        }
        
        const point = new Point2D(x, y);
        if (!point.isValid()) {
            throw new Error('Invalid point');
        }
        
        return point;
    }
}

// Base point G in extended coordinates
const G = new Point4D(GX, GY, 1n, mod(GX * GY, P));

/**
 * Modular arithmetic helper functions
 */
function mod(a, m) {
    const result = a % m;
    return result < 0n ? result + m : result;
}

function modInverse(a, m) {
    const [gcd, x] = extendedGcd(a, m);
    if (gcd !== 1n) {
        throw new Error('Modular inverse does not exist');
    }
    return mod(x, m);
}

function extendedGcd(a, b) {
    if (b === 0n) {
        return [a, 1n, 0n];
    }
    const [gcd, x1, y1] = extendedGcd(b, a % b);
    const x = y1;
    const y = x1 - (a / b) * y1;
    return [gcd, x, y];
}

function modSqrt(n, p) {
    // Tonelli-Shanks algorithm for computing square roots modulo p
    if (modPow(n, (p - 1n) / 2n, p) !== 1n) {
        return null; // n is not a quadratic residue
    }
    
    // Find Q and S such that p - 1 = Q * 2^S with Q odd
    let q = p - 1n;
    let s = 0n;
    while (q % 2n === 0n) {
        q /= 2n;
        s++;
    }
    
    if (s === 1n) {
        return modPow(n, (p + 1n) / 4n, p);
    }
    
    // Find a quadratic non-residue z
    let z = 2n;
    while (modPow(z, (p - 1n) / 2n, p) !== p - 1n) {
        z++;
    }
    
    let m = s;
    let c = modPow(z, q, p);
    let t = modPow(n, q, p);
    let r = modPow(n, (q + 1n) / 2n, p);
    
    while (t !== 1n) {
        // Find the smallest i such that t^(2^i) = 1
        let i = 1n;
        let temp = mod(t * t, p);
        while (temp !== 1n) {
            temp = mod(temp * temp, p);
            i++;
        }
        
        // Update values
        const b = modPow(c, modPow(2n, m - i - 1n, p - 1n), p);
        m = i;
        c = mod(b * b, p);
        t = mod(t * c, p);
        r = mod(r * b, p);
    }
    
    return r;
}

function modPow(base, exponent, modulus) {
    let result = 1n;
    base = mod(base, modulus);
    
    while (exponent > 0n) {
        if (exponent & 1n) {
            result = mod(result * base, modulus);
        }
        exponent >>= 1n;
        base = mod(base * base, modulus);
    }
    
    return result;
}

/**
 * Hash-to-point function H(data) -> Point
 * Used for deriving points from arbitrary data
 */
function hashToPoint(data) {
    const hash = crypto.createHash('sha512').update(data).digest();
    
    // Try to find a valid point by incrementing a counter
    for (let counter = 0; counter < 256; counter++) {
        const input = Buffer.concat([hash, Buffer.from([counter])]);
        const pointHash = crypto.createHash('sha512').update(input).digest();
        
        // Take first 32 bytes as y coordinate
        const yBytes = pointHash.slice(0, 32);
        yBytes[31] &= 0x7F; // Clear sign bit
        
        try {
            const point = Point2D.decompress(yBytes);
            return point;
        } catch {
            continue; // Try next counter value
        }
    }
    
    throw new Error('Failed to find valid point');
}

/**
 * HKDF key derivation function
 */
function hkdf(ikm, salt = null, info = null, length = 32) {
    // Extract phase
    const actualSalt = salt || Buffer.alloc(32, 0);
    const prk = crypto.createHmac('sha256', actualSalt).update(ikm).digest();
    
    // Expand phase
    const n = Math.ceil(length / 32);
    let okm = Buffer.alloc(0);
    let t = Buffer.alloc(0);
    
    for (let i = 1; i <= n; i++) {
        const hmac = crypto.createHmac('sha256', prk);
        hmac.update(t);
        if (info) hmac.update(info);
        hmac.update(Buffer.from([i]));
        t = hmac.digest();
        okm = Buffer.concat([okm, t]);
    }
    
    return okm.slice(0, length);
}

/**
 * Shamir Secret Sharing
 */
class ShamirSecretSharing {
    /**
     * Split secret into n shares with threshold t
     */
    static split(secret, threshold, numShares) {
        if (threshold > numShares) {
            throw new Error('Threshold cannot be greater than number of shares');
        }
        if (threshold < 1) {
            throw new Error('Threshold must be at least 1');
        }
        
        // For secrets larger than Q, we need to reduce them modulo Q
        // but preserve the original length for reconstruction
        const originalLength = secret.length;
        const secretInt = this.bytesToBigInt(secret) % Q;
        const coefficients = [secretInt];
        
        // Generate random coefficients for polynomial
        for (let i = 1; i < threshold; i++) {
            coefficients.push(this.randomBigInt(Q));
        }
        
        // Evaluate polynomial at points 1, 2, ..., numShares
        const shares = [];
        for (let x = 1; x <= numShares; x++) {
            let y = 0n;
            let xPower = 1n;
            
            for (const coeff of coefficients) {
                y = mod(y + mod(coeff * xPower, Q), Q);
                xPower = mod(xPower * BigInt(x), Q);
            }
            
            shares.push({ x: x, y: y, length: originalLength });
        }
        
        return shares;
    }
    
    /**
     * Recover secret from shares using Lagrange interpolation
     */
    static recover(shares) {
        if (shares.length === 0) {
            throw new Error('No shares provided');
        }
        
        let secret = 0n;
        
        for (let i = 0; i < shares.length; i++) {
            let numerator = 1n;
            let denominator = 1n;
            
            for (let j = 0; j < shares.length; j++) {
                if (i !== j) {
                    numerator = mod(numerator * BigInt(-shares[j].x), Q);
                    denominator = mod(denominator * BigInt(shares[i].x - shares[j].x), Q);
                }
            }
            
            const lagrangeCoeff = mod(numerator * modInverse(denominator, Q), Q);
            secret = mod(secret + mod(shares[i].y * lagrangeCoeff, Q), Q);
        }
        
        // Use the original length from the shares
        const originalLength = shares[0].length || 32;
        return this.bigIntToBytes(secret, originalLength);
    }
    
        static bytesToBigInt(bytes) {
        let result = 0n;
        for (let i = 0; i < bytes.length; i++) {
            result = (result << 8n) + BigInt(bytes[i]);
        }
        return result;
    }

    static bigIntToBytes(value, length) {
        const bytes = new Uint8Array(length);
        let v = value;
        for (let i = length - 1; i >= 0; i--) {
            bytes[i] = Number(v & 0xFFn);
            v >>= 8n;
        }
        return bytes;
    }
    
    static randomBigInt(max) {
        const bytes = crypto.randomBytes(32);
        let result = this.bytesToBigInt(bytes);
        return result % max;
    }
}

/**
 * Generate authentication codes for servers
 */
function generateAuthCodes(sharedSecret, serverIds) {
    const authCodes = {};
    
    for (const serverId of serverIds) {
        const input = Buffer.concat([
            Buffer.from(sharedSecret),
            Buffer.from(serverId, 'utf8')
        ]);
        const hash = crypto.createHash('sha256').update(input).digest();
        authCodes[serverId] = hash;
    }
    
    return authCodes;
}

/**
 * Convert password to PIN using PBKDF2
 */
function passwordToPin(password, salt, iterations = 100000) {
    const key = crypto.pbkdf2Sync(password, salt, iterations, 4, 'sha256');
    const pin = key.readUInt32BE(0) % 1000000;
    return pin.toString().padStart(6, '0');
}

export {
    Point2D,
    Point4D,
    G,
    P,
    Q,
    D,
    hashToPoint,
    hkdf,
    ShamirSecretSharing,
    generateAuthCodes,
    passwordToPin,
    mod,
    modInverse,
    modPow
}; 

File: sdk/cpp/src/crypto.cpp
#include "openadp/crypto.hpp"
#include "openadp/types.hpp"
#include "openadp/utils.hpp"
#include <openssl/evp.h>
#include <openssl/kdf.h>
#include <openssl/sha.h>
#include <openssl/rand.h>
#include <openssl/aes.h>
#include <openssl/ec.h>
#include <openssl/bn.h>
#include <cstring>
#include <algorithm>
#include <numeric>

namespace openadp {
namespace crypto {

// Helper function to convert hex string to BIGNUM
BIGNUM* hex_to_bn(const std::string& hex) {
    BIGNUM* bn = BN_new();
    BN_hex2bn(&bn, hex.c_str());
    return bn;
}

// Helper function to convert BIGNUM to hex string
std::string bn_to_hex(const BIGNUM* bn) {
    char* hex_str = BN_bn2hex(bn);
    std::string result(hex_str);
    OPENSSL_free(hex_str);
    return result;
}

// Hash function
Bytes sha256_hash(const Bytes& data) {
    Bytes result(SHA256_DIGEST_LENGTH);
    SHA256(data.data(), data.size(), result.data());
    return result;
}

// Prefixed function (length prefix + data)
Bytes prefixed(const Bytes& data) {
    Bytes result;
    uint16_t length = static_cast<uint16_t>(data.size());
    
    // Little-endian encoding (16-bit)
    result.push_back(length & 0xFF);
    result.push_back((length >> 8) & 0xFF);
    
    result.insert(result.end(), data.begin(), data.end());
    return result;
}

std::string bytes_to_hex(const Bytes& data) {
    return utils::hex_encode(data);
}

Bytes hex_to_bytes(const std::string& hex) {
    return utils::hex_decode(hex);
}

// Ed25519 implementation
Point4D Ed25519::hash_to_point(const Bytes& uid, const Bytes& did, const Bytes& bid, const Bytes& pin) {
    // Concatenate all inputs with length prefixes
    Bytes prefixed_uid = prefixed(uid);
    Bytes prefixed_did = prefixed(did);
    Bytes prefixed_bid = prefixed(bid);
    
    Bytes data;
    data.insert(data.end(), prefixed_uid.begin(), prefixed_uid.end());
    data.insert(data.end(), prefixed_did.begin(), prefixed_did.end());
    data.insert(data.end(), prefixed_bid.begin(), prefixed_bid.end());
    data.insert(data.end(), pin.begin(), pin.end());
    
    // Hash and convert to point
    Bytes hash_bytes = sha256_hash(data);
    
    // Convert hash to big integer (little-endian)
    BIGNUM* y_base = BN_new();
    BN_zero(y_base);
    
    for (size_t i = 0; i < hash_bytes.size(); ++i) {
        BIGNUM* byte_bn = BN_new();
        BN_set_word(byte_bn, hash_bytes[i]);
        
        BIGNUM* shift_bn = BN_new();
        BN_set_word(shift_bn, 1);
        BN_lshift(shift_bn, shift_bn, i * 8);
        
        BN_mul(byte_bn, byte_bn, shift_bn, BN_CTX_new());
        BN_add(y_base, y_base, byte_bn);
        
        BN_free(byte_bn);
        BN_free(shift_bn);
    }
    
    // Reduce modulo curve order
    BIGNUM* curve_order = hex_to_bn("1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed");
    BN_mod(y_base, y_base, curve_order, BN_CTX_new());
    
    // Create point coordinates (simplified - this would need proper Ed25519 point construction)
    std::string y_hex = bn_to_hex(y_base);
    
    BN_free(y_base);
    BN_free(curve_order);
    
    // For now, return a simplified point (this needs proper Ed25519 implementation)
    return Point4D("0", y_hex, "1", "0");
}

Point4D Ed25519::scalar_mult(const std::string& scalar_hex, const Point4D& point) {
    // This is a simplified implementation - would need proper Ed25519 scalar multiplication
    BIGNUM* scalar = hex_to_bn(scalar_hex);
    BIGNUM* y = hex_to_bn(point.y);
    
    // Simplified multiplication (not cryptographically correct)
    BN_mul(y, y, scalar, BN_CTX_new());
    
    std::string result_y = bn_to_hex(y);
    
    BN_free(scalar);
    BN_free(y);
    
    return Point4D(point.x, result_y, point.z, point.t);
}

Point4D Ed25519::point_add(const Point4D& p1, const Point4D& p2) {
    // Simplified point addition (not cryptographically correct)
    BIGNUM* y1 = hex_to_bn(p1.y);
    BIGNUM* y2 = hex_to_bn(p2.y);
    
    BN_add(y1, y1, y2);
    
    std::string result_y = bn_to_hex(y1);
    
    BN_free(y1);
    BN_free(y2);
    
    return Point4D("0", result_y, "1", "0");
}

Bytes Ed25519::compress(const Point4D& point) {
    // Convert point to compressed form
    Bytes result = hex_to_bytes(point.y);
    if (result.size() < 32) {
        result.resize(32, 0);
    }
    return result;
}

Point4D Ed25519::decompress(const Bytes& data) {
    std::string y_hex = bytes_to_hex(data);
    return Point4D("0", y_hex, "1", "0");
}

Point4D Ed25519::expand(const Point2D& point) {
    return Point4D(point.x, point.y, "1", "0");
}

Point2D Ed25519::unexpand(const Point4D& point) {
    return Point2D(point.x, point.y);
}

bool Ed25519::is_valid_point(const Point4D& point) {
    // Simplified validation
    return !point.y.empty();
}

Point4D Ed25519::point_mul8(const Point4D& point) {
    return scalar_mult("8", point);
}

// Shamir Secret Sharing
std::vector<Share> ShamirSecretSharing::split_secret(const std::string& secret_hex, int threshold, int num_shares) {
    BIGNUM* secret = hex_to_bn(secret_hex);
    BIGNUM* prime = hex_to_bn("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F"); // secp256k1 prime
    
    // Generate random coefficients
    std::vector<BIGNUM*> coefficients;
    coefficients.push_back(BN_dup(secret)); // a0 = secret
    
    for (int i = 1; i < threshold; i++) {
        BIGNUM* coeff = BN_new();
        BN_rand_range(coeff, prime);
        coefficients.push_back(coeff);
    }
    
    // Generate shares
    std::vector<Share> shares;
    for (int x = 1; x <= num_shares; x++) {
        BIGNUM* y = BN_new();
        BN_zero(y);
        
        BIGNUM* x_power = BN_new();
        BN_one(x_power);
        
        for (int i = 0; i < threshold; i++) {
            BIGNUM* term = BN_new();
            BN_mul(term, coefficients[i], x_power, BN_CTX_new());
            BN_add(y, y, term);
            
            BN_mul_word(x_power, x);
            BN_free(term);
        }
        
        BN_mod(y, y, prime, BN_CTX_new());
        shares.emplace_back(x, bn_to_hex(y));
        
        BN_free(y);
        BN_free(x_power);
    }
    
    // Cleanup
    for (BIGNUM* coeff : coefficients) {
        BN_free(coeff);
    }
    BN_free(secret);
    BN_free(prime);
    
    return shares;
}

std::string ShamirSecretSharing::recover_secret(const std::vector<Share>& shares) {
    if (shares.empty()) {
        throw OpenADPError("No shares provided");
    }
    
    BIGNUM* prime = hex_to_bn("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F");
    BIGNUM* result = BN_new();
    BN_zero(result);
    
    // Lagrange interpolation
    for (size_t i = 0; i < shares.size(); i++) {
        BIGNUM* numerator = BN_new();
        BIGNUM* denominator = BN_new();
        BN_one(numerator);
        BN_one(denominator);
        
        for (size_t j = 0; j < shares.size(); j++) {
            if (i != j) {
                // Numerator: multiply by shares[j].x
                BN_mul_word(numerator, shares[j].x);
                
                // Denominator: multiply by (shares[j].x - shares[i].x)
                BIGNUM* diff = BN_new();
                if (shares[j].x >= shares[i].x) {
                    BN_set_word(diff, shares[j].x - shares[i].x);
                } else {
                    // Handle negative difference: compute prime - (shares[i].x - shares[j].x)
                    BN_set_word(diff, shares[i].x - shares[j].x);
                    BIGNUM* temp = BN_dup(prime);
                    BN_sub(temp, temp, diff);
                    BN_copy(diff, temp);
                    BN_free(temp);
                }
                
                BN_mul(denominator, denominator, diff, BN_CTX_new());
                BN_free(diff);
            }
        }
        
        // Modular inverse
        BIGNUM* inv = BN_new();
        BN_mod_inverse(inv, denominator, prime, BN_CTX_new());
        
        BIGNUM* lagrange = BN_new();
        BN_mul(lagrange, numerator, inv, BN_CTX_new());
        BN_mod(lagrange, lagrange, prime, BN_CTX_new());
        
        BIGNUM* y = hex_to_bn(shares[i].y);
        BN_mul(lagrange, lagrange, y, BN_CTX_new());
        BN_add(result, result, lagrange);
        
        BN_free(numerator);
        BN_free(denominator);
        BN_free(inv);
        BN_free(lagrange);
        BN_free(y);
    }
    
    BN_mod(result, result, prime, BN_CTX_new());
    std::string secret_hex = bn_to_hex(result);
    
    BN_free(result);
    BN_free(prime);
    
    return secret_hex;
}

// Point Secret Sharing (simplified)
std::vector<PointShare> PointSecretSharing::split_point(const Point2D& point, int threshold, int num_shares) {
    auto x_shares = ShamirSecretSharing::split_secret(point.x, threshold, num_shares);
    auto y_shares = ShamirSecretSharing::split_secret(point.y, threshold, num_shares);
    
    std::vector<PointShare> point_shares;
    for (int i = 0; i < num_shares; i++) {
        Point2D share_point(x_shares[i].y, y_shares[i].y);
        point_shares.emplace_back(x_shares[i].x, share_point);
    }
    
    return point_shares;
}

Point2D PointSecretSharing::recover_point(const std::vector<PointShare>& shares) {
    std::vector<Share> x_shares, y_shares;
    
    for (const auto& point_share : shares) {
        x_shares.emplace_back(point_share.x, point_share.point.x);
        y_shares.emplace_back(point_share.x, point_share.point.y);
    }
    
    std::string x = ShamirSecretSharing::recover_secret(x_shares);
    std::string y = ShamirSecretSharing::recover_secret(y_shares);
    
    return Point2D(x, y);
}

// Key derivation
Bytes derive_encryption_key(const Point4D& point) {
    Bytes point_bytes = Ed25519::compress(point);
    return sha256_hash(point_bytes);
}

// AES-GCM encryption
AESGCMResult aes_gcm_encrypt(const Bytes& plaintext, const Bytes& key, const Bytes& associated_data) {
    if (key.size() != 32) {
        throw OpenADPError("AES key must be 32 bytes");
    }
    
    // Generate random nonce
    Bytes nonce = utils::random_bytes(12);
    
    EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
    if (!ctx) {
        throw OpenADPError("Failed to create cipher context");
    }
    
    // Initialize encryption
    if (EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to initialize AES-GCM");
    }
    
    // Set nonce length
    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, nonce.size(), nullptr) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to set nonce length");
    }
    
    // Set key and nonce
    if (EVP_EncryptInit_ex(ctx, nullptr, nullptr, key.data(), nonce.data()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to set key and nonce");
    }
    
    // Set associated data
    int len;
    if (!associated_data.empty()) {
        if (EVP_EncryptUpdate(ctx, nullptr, &len, associated_data.data(), associated_data.size()) != 1) {
            EVP_CIPHER_CTX_free(ctx);
            throw OpenADPError("Failed to set associated data");
        }
    }
    
    // Encrypt
    Bytes ciphertext(plaintext.size());
    if (EVP_EncryptUpdate(ctx, ciphertext.data(), &len, plaintext.data(), plaintext.size()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Encryption failed");
    }
    ciphertext.resize(len);
    
    // Finalize
    Bytes final_block(16);
    if (EVP_EncryptFinal_ex(ctx, final_block.data(), &len) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Encryption finalization failed");
    }
    
    if (len > 0) {
        ciphertext.insert(ciphertext.end(), final_block.begin(), final_block.begin() + len);
    }
    
    // Get tag
    Bytes tag(16);
    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, 16, tag.data()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to get authentication tag");
    }
    
    EVP_CIPHER_CTX_free(ctx);
    
    return AESGCMResult{ciphertext, tag, nonce};
}

AESGCMResult aes_gcm_encrypt(const Bytes& plaintext, const Bytes& key) {
    return aes_gcm_encrypt(plaintext, key, Bytes{});
}

Bytes aes_gcm_decrypt(const Bytes& ciphertext, const Bytes& tag, const Bytes& nonce, 
                     const Bytes& key, const Bytes& associated_data) {
    if (key.size() != 32) {
        throw OpenADPError("AES key must be 32 bytes");
    }
    
    EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
    if (!ctx) {
        throw OpenADPError("Failed to create cipher context");
    }
    
    // Initialize decryption
    if (EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to initialize AES-GCM");
    }
    
    // Set nonce length
    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, nonce.size(), nullptr) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to set nonce length");
    }
    
    // Set key and nonce
    if (EVP_DecryptInit_ex(ctx, nullptr, nullptr, key.data(), nonce.data()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to set key and nonce");
    }
    
    // Set associated data
    int len;
    if (!associated_data.empty()) {
        if (EVP_DecryptUpdate(ctx, nullptr, &len, associated_data.data(), associated_data.size()) != 1) {
            EVP_CIPHER_CTX_free(ctx);
            throw OpenADPError("Failed to set associated data");
        }
    }
    
    // Decrypt
    Bytes plaintext(ciphertext.size());
    if (EVP_DecryptUpdate(ctx, plaintext.data(), &len, ciphertext.data(), ciphertext.size()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Decryption failed");
    }
    plaintext.resize(len);
    
    // Set tag
    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, tag.size(), const_cast<uint8_t*>(tag.data())) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw OpenADPError("Failed to set authentication tag");
    }
    
    // Finalize
    Bytes final_block(16);
    int ret = EVP_DecryptFinal_ex(ctx, final_block.data(), &len);
    EVP_CIPHER_CTX_free(ctx);
    
    if (ret != 1) {
        throw OpenADPError("Authentication tag verification failed");
    }
    
    if (len > 0) {
        plaintext.insert(plaintext.end(), final_block.begin(), final_block.begin() + len);
    }
    
    return plaintext;
}

Bytes aes_gcm_decrypt(const Bytes& ciphertext, const Bytes& tag, const Bytes& nonce, 
                     const Bytes& key) {
    return aes_gcm_decrypt(ciphertext, tag, nonce, key, Bytes{});
}

// HKDF key derivation
Bytes hkdf_derive(const Bytes& input_key, const Bytes& salt, const Bytes& info, size_t output_length) {
    EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, nullptr);
    if (!ctx) {
        throw OpenADPError("Failed to create HKDF context");
    }
    
    if (EVP_PKEY_derive_init(ctx) != 1) {
        EVP_PKEY_CTX_free(ctx);
        throw OpenADPError("Failed to initialize HKDF");
    }
    
    if (EVP_PKEY_CTX_set_hkdf_md(ctx, EVP_sha256()) != 1) {
        EVP_PKEY_CTX_free(ctx);
        throw OpenADPError("Failed to set HKDF hash function");
    }
    
    if (EVP_PKEY_CTX_set1_hkdf_key(ctx, input_key.data(), input_key.size()) != 1) {
        EVP_PKEY_CTX_free(ctx);
        throw OpenADPError("Failed to set HKDF input key");
    }
    
    if (!salt.empty()) {
        if (EVP_PKEY_CTX_set1_hkdf_salt(ctx, salt.data(), salt.size()) != 1) {
            EVP_PKEY_CTX_free(ctx);
            throw OpenADPError("Failed to set HKDF salt");
        }
    }
    
    if (!info.empty()) {
        if (EVP_PKEY_CTX_add1_hkdf_info(ctx, info.data(), info.size()) != 1) {
            EVP_PKEY_CTX_free(ctx);
            throw OpenADPError("Failed to set HKDF info");
        }
    }
    
    Bytes output(output_length);
    size_t out_len = output_length;
    
    if (EVP_PKEY_derive(ctx, output.data(), &out_len) != 1) {
        EVP_PKEY_CTX_free(ctx);
        throw OpenADPError("HKDF derivation failed");
    }
    
    EVP_PKEY_CTX_free(ctx);
    output.resize(out_len);
    
    return output;
}

} // namespace crypto
} // namespace openadp 